### ISSUE TITLE
WIP: Training MNIST based Linear model e2e

### ISSUE THREAD

**Issue by nvukobratTT** [tenstorrent/tt-mlir#156]
### Summary
In order to train this model, the following key details are required:
- Required `fwd`, `bwd`, `loss` and `opt` ops are supported e2e (#77 #145)
- Inference is running e2e (#149)
- Runtime API can handle CPU/TT model execution (#154)

Beneficial requirements are:
- We have valid pre-trained weights (#150)
- Training infrastructure is setup (#151)

Having the above-mentioned components, pushing initial e2e training should be straightforward.

### Details
With all the mentioned components, most of the details should be in place. For easier understanding, here is the overview of how compile flow should work and look like:
1. We define our training model and training loop. Check `pybuda/test/mlir/mnist/test_training.py` for a sample. 

### Update how model is initialized
Provided example should be update from this:
```py
    framework_model = MNISTLinear()
    tt_model, tt_optimizer, tt_loss = pybuda.compile(
        pybuda.PyTorchModule("mnist_linear", framework_model),
        optimizer=torch.optim.SGD(framework_model.parameters(), lr=learning_rate),
        loss_fn=torch.nn.L1Loss(),
        backend="tt",
    )
```

to something like this:
```py
    # Initialize models
    framework_model = MNISTLinear()
    framework_optimizer = torch.optim.SGD(framework_model.parameters(), lr=learning_rate)
    framework_loss = torch.nn.L1Loss()
    
    # Compile models
    tt_model = pybuda.compile(framework_model.to("tt"))
    tt_optimizer = pybuda.compile(framework_optimizer.to("tt"))
    tt_loss = pybuda.compile(framework_loss.to("tt"))
```

### Limitations
Each limitation represents core concepts that we need for scalable e2e training. Each of these limitations will be tackled through separate issues. 

Limitations:
- Training with recompute will not work for the initial version
    - Initial training will not include training with recompute. 
    - Main point is that MLIR should be as agnostic as possible to the ML. I.e. it should only see different graphs that should be run. E.g. forwards, backward, optimizer. 
    - Also, the new BE is a temporal-based stack, so older implementation will not be easily applicable from spatial to temporal atp.
    - Therefore, we can't keep the concept of interconnected fwd and bwd ops in order to support recompute as done before. So, new solution should be introduced that will support both PyBuda and Jax frontends. TBD.
- Multichip will not work for the initial version
    - Training on multichip has additional caveats 
- Not mandatory to be efficient from the start
    - For further experiments, we need to focus on multi-chip and larger models. E.g. for valid PoC, we need to demonstrate that we can scale the training across multiple machines in order to be able to train larger models

### Passing Criteria

**Comment by nsmithtt** [comment]
I think this section should be edited:
```python
    tt_model = pybuda.compile(framework_model.to("tt"))
    tt_optimizer = pybuda.compile(framework_optimizer.to("tt"))
    tt_loss = pybuda.compile(framework_loss.to("tt"))
```

To just:
```python
    tt_model = pybuda.compile(framework_model)
    tt_optimizer = pybuda.compile(framework_optimizer)
    tt_loss = pybuda.compile(framework_loss)
```

And then when the respective module is ready to be run on the device you have to do:
```python
tt_model.to("tt")
```

What `.to("tt")` does on a module is prepare/copy the weights to device DRAM.  It could be legal to run:
```python
tt_model.to("tt")
tt_loss.to("tt")
```

But what this means is that all of the parameters for `tt_model` and `tt_loss` will be copied to device DRAM and live simultaneously there.

**Comment by nsmithtt** [comment]
Nikola, this looks great!  Once we have a POC it'd be great to turn this into a documentation spec for training.

**Comment by nsmithtt** [comment]
I think an R&D effort needs to try and see how we can hook:
```python
loss.backward() # executes bwd graph
tt_optimizer.step() # executes optimizer graph
```

An additional challenge is that if loss is separately compiled regardless of it being scheduled on CPU or TT, is how we can propagate gradients between graphs.  At first glance it seems like it'd just be a matter of wiring up loss backward outputs -> model backward inputs, but there are likely some finicky details to work out.  For example, when `loss.backward()` is called it implicitly runs `tt_model` backward, so I think the runtime will need to keep track of this and schedule this automatically.

Optimizer step is essentially the same issue, but it feels slightly more explicit since the outputs are given up front and it somewhat decoupled from the loss and model graphs.

A potentially reasonable intermediate step could be to force users to embed loss into their module directly:

```python
class ModuleAndLoss(nn.Module):
    def __init__(self, module, loss):
        self.module = module
        self.loss = loss

    def forward(self, inputs, targets):
        return self.loss(self.module(inputs), targets)
```

This has the benefit of not having to propagate gradients through flatbuffer boundaries, however, in the long term that would provide most flexibility.

**Comment by nvukobratTT** [comment]
> details regarding `.to("tt")`
Similar as on [this](https://github.com/tenstorrent/tt-mlir/issues/154) discussion, I agree! 

> Nikola, this looks great! Once we have a POC it'd be great to turn this into a documentation spec for training.
For sure! 

> An additional challenge is that if loss is separately compiled regardless of it being scheduled on CPU or TT, is how we can propagate gradients between graphs. At first glance it seems like it'd just be a matter of wiring up loss backward outputs -> model backward inputs, but there are likely some finicky details to work out. For example, when loss.backward() is called it implicitly runs tt_model backward, so I think the runtime will need to keep track of this and schedule this automatically.

I see your reasoning. The caveat of `backward()` is not outputs propagation, but rather runtime awareness how that should be mapped. More precisely, when we call `loss.backward()`, runtime needs to:
1. Invoke loss bwd and store gradients on desired location
2. Invoke `tt_model` bwd but fetching loss bwd output, and running rest of the bwd graph
3. Store full bwd output to the desired location

Regarding optimizer, it's a bit more simpler, as it doesn't need to run different modules. However, it has to be aware of where accumulated gradients are stored, and where are parameters that needs to be updated. 

Let's see during R&D what are the tricky points.
